File: context_shield/policy.py
"""PolicyEngine: loads and evaluates trust-based access control rules."""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

import yaml
from pydantic import ValidationError

from context_shield.exceptions import PolicyLoadError, PolicyValidationError
from context_shield.types import (
    Channel,
    PolicyConfig,
    PolicyRule,
    TrustLevel,
)

logger = logging.getLogger(__name__)


class PolicyEngine:
    """Loads and evaluates policy rules against (channel, trust, tool_name) tuples.

    Evaluation uses "last match wins with explicit priority":
    1. Collect all rules matching (channel, trust) — wildcards included
    2. Sort by priority ascending
    3. Walk sorted list; each matching deny/allow updates the verdict
    4. Final verdict stands. If no rule ever mentioned the tool, default_deny applies.
    """

    def __init__(self, config: PolicyConfig) -> None:
        self._config = config
        self._rules = sorted(config.rules, key=lambda r: r.priority)

    @classmethod
    def from_yaml(cls, path: Path | str) -> PolicyEngine:
        """Load policy from a YAML file."""
        path = Path(path)
        try:
            raw = path.read_text(encoding="utf-8")
        except OSError as e:
            raise PolicyLoadError(f"Cannot read policy file: {path}") from e

        try:
            data = yaml.safe_load(raw)
        except yaml.YAMLError as e:
            raise PolicyLoadError(f"Invalid YAML in {path}") from e

        return cls.from_dict(data)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> PolicyEngine:
        """Load policy from a dict."""
        try:
            config = PolicyConfig(**data)
        except (ValidationError, TypeError) as e:
            raise PolicyValidationError(f"Invalid policy config: {e}") from e
        return cls(config)

    def reload(self, path: Path | str) -> None:
        """Hot-reload policy from file."""
        engine = self.from_yaml(path)
        self._config = engine._config
        self._rules = engine._rules
        logger.info("Policy reloaded from %s", path)

    def evaluate(
        self,
        tool_name: str,
        channel: Channel,
        trust: TrustLevel,
    ) -> tuple[bool, PolicyRule | None, str]:
        """Evaluate whether a tool call is permitted.

        Returns (allowed, matched_rule, reason).
        """
        verdict: bool | None = None
        matched_rule: PolicyRule | None = None

        for rule in self._rules:
            if not self._matches_rule(rule, channel, trust):
                continue

            if tool_name in rule.deny or "*" in rule.deny:
                verdict = False
                matched_rule = rule
            if tool_name in rule.allow or "*" in rule.allow:
                verdict = True
                matched_rule = rule

        if verdict is None:
            if self._config.default_deny:
                return (
                    False,
                    None,
                    f"No rule matched for tool '{tool_name}' on "
                    f"channel={channel.value} trust={trust.name}; "
                    f"default policy is deny",
                )
            return (
                True,
                None,
                f"No rule matched for tool '{tool_name}'; default policy is allow",
            )

        action = "allowed" if verdict else "denied"
        reason = (
            f"Tool '{tool_name}' {action} by rule "
            f"(channel={matched_rule.channel}, trust={matched_rule.trust}, "
            f"priority={matched_rule.priority})"
        )
        return verdict, matched_rule, reason

    def _matches_rule(
        self, rule: PolicyRule, channel: Channel, trust: TrustLevel
    ) -> bool:
        """Check if a rule's selectors match the given channel and trust."""
        channel_match = rule.channel == "*" or rule.channel == channel
        trust_match = rule.trust == "*" or rule.trust == trust
        return channel_match and trust_match

    def validate(self) -> list[str]:
        """Check policy for potential misconfigurations. Returns warnings."""
        warnings: list[str] = []
        for i, rule in enumerate(self._rules):
            if "*" in rule.allow and rule.trust in (
                TrustLevel.UNTRUSTED,
                TrustLevel.HOSTILE,
                "*",
            ):
                warnings.append(
                    f"Rule {i}: wildcard allow on "
                    f"trust={rule.trust} — this may be overly permissive"
                )
            if rule.allow and rule.deny:
                overlap = set(rule.allow) & set(rule.deny)
                if overlap:
                    warnings.append(
                        f"Rule {i}: tools {overlap} appear in both "
                        f"allow and deny — deny checked first, "
                        f"then allow overrides"
                    )
        return warnings

    @property
    def config(self) -> PolicyConfig:
        return self._config

File: rust/intake-core/src/policy.rs
//! PolicyEngine: loads and evaluates trust-based access control rules.
//!
//! Direct port of `intake_shield/policy.py`.
//!
//! Evaluation uses "last match wins with explicit priority":
//! 1. Collect all rules matching (channel, trust) — wildcards included
//! 2. Sort by priority ascending
//! 3. Walk sorted list; each matching deny/allow updates the verdict
//! 4. Final verdict stands. If no rule ever mentioned the tool, default_deny applies.

use std::path::Path;

use crate::error::IntakeError;
use crate::types::{Channel, EvalResult, Matcher, PolicyConfig, PolicyRule, TrustLevel};

/// Loads and evaluates policy rules against `(channel, trust, tool_name)` tuples.
pub struct PolicyEngine {
    config: PolicyConfig,
    /// Rules pre-sorted by priority ascending.
    rules: Vec<PolicyRule>,
}

impl PolicyEngine {
    /// Create from an already-parsed config.
    pub fn from_config(config: PolicyConfig) -> Self {
        let mut rules = config.rules.clone();
        rules.sort_by_key(|r| r.priority);
        Self { config, rules }
    }

    /// Load policy from a YAML file.
    pub fn from_yaml(path: &Path) -> Result<Self, IntakeError> {
        let raw = std::fs::read_to_string(path).map_err(|e| {
            IntakeError::PolicyLoad(format!("Cannot read policy file: {}: {e}", path.display()))
        })?;
        Self::parse_yaml(&raw)
    }

    /// Load policy from a YAML string.
    pub fn parse_yaml(yaml: &str) -> Result<Self, IntakeError> {
        let config: PolicyConfig =
            serde_yaml::from_str(yaml).map_err(|e| IntakeError::PolicyValidation(e.to_string()))?;
        Ok(Self::from_config(config))
    }

    /// Load policy from a dict-like structure (serde_json::Value or similar).
    pub fn from_value(value: serde_json::Value) -> Result<Self, IntakeError> {
        let config: PolicyConfig = serde_json::from_value(value)
            .map_err(|e| IntakeError::PolicyValidation(e.to_string()))?;
        Ok(Self::from_config(config))
    }

    /// Hot-reload policy from file.
    pub fn reload(&mut self, path: &Path) -> Result<(), IntakeError> {
        let new = Self::from_yaml(path)?;
        self.config = new.config;
        self.rules = new.rules;
        Ok(())
    }

    /// Evaluate whether a tool call is permitted.
    ///
    /// Returns `EvalResult { allowed, matched_rule, reason }`.
    pub fn evaluate(
        &self,
        tool_name: &str,
        channel: Channel,
        trust: TrustLevel,
    ) -> EvalResult {
        let mut verdict: Option<bool> = None;
        let mut matched_rule: Option<&PolicyRule> = None;

        for rule in &self.rules {
            if !Self::matches_rule(rule, channel, trust) {
                continue;
            }

            if rule.deny.contains(&tool_name.to_string())
                || rule.deny.iter().any(|d| d == "*")
            {
                verdict = Some(false);
                matched_rule = Some(rule);
            }
            if rule.allow.contains(&tool_name.to_string())
                || rule.allow.iter().any(|a| a == "*")
            {
                verdict = Some(true);
                matched_rule = Some(rule);
            }
        }

        match verdict {
            None => {
                if self.config.default_deny {
                    EvalResult {
                        allowed: false,
                        matched_rule: None,
                        reason: format!(
                            "No rule matched for tool '{tool_name}' on channel={channel} trust={trust}; default policy is deny"
                        ),
                    }
                } else {
                    EvalResult {
                        allowed: true,
                        matched_rule: None,
                        reason: format!(
                            "No rule matched for tool '{tool_name}'; default policy is allow"
                        ),
                    }
                }
            }
            Some(allowed) => {
                let rule = matched_rule.expect("matched_rule must be set when verdict is Some");
                let action = if allowed { "allowed" } else { "denied" };
                let reason = format!(
                    "Tool '{}' {} by rule (channel={}, trust={}, priority={})",
                    tool_name,
                    action,
                    match &rule.channel {
                        Matcher::Wildcard => "*".to_string(),
                        Matcher::Specific(c) => c.to_string(),
                    },
                    match &rule.trust {
                        Matcher::Wildcard => "*".to_string(),
                        Matcher::Specific(t) => t.to_string(),
                    },
                    rule.priority,
                );
                EvalResult {
                    allowed,
                    matched_rule: Some(rule.clone()),
                    reason,
                }
            }
        }
    }

    /// Check policy for potential misconfigurations. Returns warnings.
    pub fn validate(&self) -> Vec<String> {
        let mut warnings = Vec::new();

        for (i, rule) in self.rules.iter().enumerate() {
            // Warn on wildcard allow with untrusted/hostile trust
            if rule.allow.iter().any(|a| a == "*") {
                let is_risky = matches!(
                    &rule.trust,
                    Matcher::Specific(TrustLevel::Untrusted)
                        | Matcher::Specific(TrustLevel::Hostile)
                        | Matcher::Wildcard
                );
                if is_risky {
                    warnings.push(format!(
                        "Rule {i}: wildcard allow on trust={} — this may be overly permissive",
                        match &rule.trust {
                            Matcher::Wildcard => "*".to_string(),
                            Matcher::Specific(t) => t.to_string(),
                        }
                    ));
                }
            }

            // Warn on tools in both allow and deny
            if !rule.allow.is_empty() && !rule.deny.is_empty() {
                let allow_set: std::collections::HashSet<_> = rule.allow.iter().collect();
                let deny_set: std::collections::HashSet<_> = rule.deny.iter().collect();
                let overlap: Vec<_> = allow_set.intersection(&deny_set).collect();
                if !overlap.is_empty() {
                    let overlap_str: Vec<_> = overlap.iter().map(|s| s.as_str()).collect();
                    warnings.push(format!(
                        "Rule {i}: tools {{{}}} appear in both allow and deny — deny checked first, then allow overrides",
                        overlap_str.join(", ")
                    ));
                }
            }
        }

        warnings
    }

    /// Access the underlying config.
    pub fn config(&self) -> &PolicyConfig {
        &self.config
    }

    fn matches_rule(rule: &PolicyRule, channel: Channel, trust: TrustLevel) -> bool {
        rule.channel.matches(&channel) && rule.trust.matches(&trust)
    }
}
